src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java

```diff
@@ -311,6 +324,83 @@ private CompletableFuture<Void> submitInstallAsync(EntandoBundleJobEntity parent
+
+                List<ComponentUsage> componentUsageList = getComponentUsageList(componentJobQueueDiff);
+
+                JobScheduler schedulerDiff = new JobScheduler();
+                schedulerDiff.queuePrimaryComponents(componentJobQueueDiff);
+                String bundleRootFolder = BundleUtilities.composeBundleResourceRootFolter(latestBundleReader);
+
+                return uninstallDiff(bundleRootFolder, schedulerDiff, componentUsageList);
+
+            } else {
+                String warnings = "The uninstallation of orphaned components is not necessary since the bundle is not installed on the system";
+                log.debug(warnings);
+                return warnings;
+            }
+        } catch (Exception e) {
+            String error = "The uninstallation of orphaned components failed with error " + e;
+            log.error(error);
+            return error;
+        }
+
+    }
+
+    private List<ComponentUsage> getComponentUsageList(Queue<EntandoBundleComponentJobEntity> componentJobQueueDiff) {
+        List<EntandoBundleComponentJobEntity> componentJobListDiff = new ArrayList<>(componentJobQueueDiff);
+        return usageService.getComponentsUsageDetails(componentJobListDiff).stream()
+                .filter(componentUsage -> {
+                    if (!componentUsage.isExist() || componentUsage.getHasExternal()) {
+                        log.debug(String.format(
+                                "Component '%s' is not found or contains external references so it can't be uninstalled",
+                                componentUsage.getCode()));
+                        return false;
+                    }
+                    return true;
+                }).collect(Collectors.toList());
+    }
+
+    private static Queue<EntandoBundleComponentJobEntity> getBundleComponentJobEntityQueue(EntandoBundleJobEntity parentJob,
+                                                                                           Queue<Installable> bundleInstallableComponentsDiff) {
+        return bundleInstallableComponentsDiff.stream()
+                .filter(i -> {
+                    if (i.getRepresentation() instanceof WidgetDescriptor) {
+                        WidgetDescriptor wd = (WidgetDescriptor) i.getRepresentation();
+                        return TYPE_WIDGET_APPBUILDER.equals(wd.getType());
+                    }
+                    return false;
+                })
+                .map(i -> {
+                    EntandoBundleComponentJobEntity cj = new EntandoBundleComponentJobEntity();
+                    cj.setParentJob(parentJob);
+                    cj.setComponentType(i.getComponentType());
+                    cj.setComponentId(i.getName());
+                    cj.setChecksum(i.getChecksum());
+                    cj.setInstallable(i);
+                    cj.setAction(i.getAction());
+                    return cj;
```

Comment:
This code should be already present in the same file, please centralize it to a designated method.